VEP22: ChangedBlockTracking: enable add/remove of qcow2 overlay if vm matches label selector

ShellyKa13 opened the pull request:
This is the first PR for the implementation of [VEP](https://github.com/kubevirt/enhancements/blob/main/veps/sig-storage/incremental-backup.md)  for incremental backup.
In this PR we are adding the API that enables doing changedblocktracking for a vm. This capability is mandatory for incremental backup.
In the VEP there is the full design of the feature. 
In short in this PR Ive added to Kubevirt CR a label selector configuration that vms that match this selector or vm that resides in a namespace that matches this selector will be initialized with CBT.
Added a state to reflect that process.
Currently restart is required to enable the changes and the added state reflects that.
The process includes creating a backend storage pvc (if doesnt already exist) creating a qcow2 overlay for the raw image and update the domain xml to reflect that.

Jira-ticket: https://issues.redhat.com/browse/CNV-61544
/label approved-vep

### Release note
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
ChangedBlockTracking: enable add/remove of qcow2 overlay if vm matches label selector
```

alicefr commented on pkg/storage/backup/cbt.go:
```diff
@@ -0,0 +1,166 @@
+		vm.Status.ChangedBlockTracking = v1.ChangedBlockTrackingEnabled
+	default:
+		log.Log.Object(vm).Warning("invalid changedBlockTracking state, removing state")
+		vm.Status.ChangedBlockTracking = ""
```
can we use a well defined status here?

alicefr commented on pkg/storage/backup/cbt.go:
```diff
@@ -0,0 +1,166 @@
+		}
+	default:
+		log.Log.Object(vm).Warning("invalid changedBlockTracking state, removing state")
+		vm.Status.ChangedBlockTracking = ""
```
same here

alicefr commented on pkg/storage/backup/cbt.go:
```diff
@@ -0,0 +1,166 @@
+/*
```
We are missing unit tests for this package

ShellyKa13 replied:
Its in the last commit

alicefr commented:
@ShellyKa13 can you put a brief description how you are going to require the backend storage with this change? Is it for every VM?

alicefr commented on pkg/virt-launcher/virtwrap/converter/converter.go:
```diff
@@ -696,6 +709,26 @@ func Convert_v1_FilesystemVolumeSource_To_api_Disk(volumeName string, disk *api.
+func Convert_v1_FilesystemVolumeSource_WithCBT_To_api_Disk(volumeName, cbtPath string, disk *api.Disk, volumesDiscardIgnore []string) error {
+	disk.Type = "file"
+	disk.Driver.Type = "qcow2"
+	disk.Driver.ErrorPolicy = v1.DiskErrorPolicyStop
```
does the error policy depends on the original one set on the disk?

alicefr commented on pkg/virt-launcher/virtwrap/manager.go:
```diff
@@ -1194,6 +1212,110 @@ func (l *LibvirtDomainManager) SyncVMI(vmi *v1.VirtualMachineInstance, allowEmul
+	}
+	overlaySize := info.VirtualSize
+
+	qmpCapabilities := `{"execute": "qmp_capabilities"}`
+	blockdevCreate := fmt.Sprintf(`{"execute": "blockdev-create", "arguments": {"job-id": "create", "options": {"driver": "qcow2", "file": "file", "data-file": "data-file", "data-file-raw": true, "size": %d}}}`, overlaySize)
+	jobDismiss := `{"execute": "job-dismiss", "arguments": {"id": "create"}}`
+	quit := `{"execute": "quit"}`
+	cmdInput := fmt.Sprintf("%s\n%s\n%s\n%s\n", qmpCapabilities, blockdevCreate, jobDismiss, quit)
+
+	args := append([]string{},
+		"--chardev", "stdio,id=stdio", "--monitor", "stdio",
+		"--blockdev", fmt.Sprintf("file,node-name=file,filename=%s", overlayPath))
```
isn't any other way rather then using directly hardcoded QMP protocol?

Acedus commented on pkg/virt-launcher/virtwrap/manager.go:
```diff
@@ -1194,6 +1212,110 @@ func (l *LibvirtDomainManager) SyncVMI(vmi *v1.VirtualMachineInstance, allowEmul
 
+var createQCOW2Overlay = createQCOW2OverlayFunc
+
+func createQCOW2OverlayFunc(overlayPath, imagePath string, blockDev bool) error {
```
Can we use the `qemu-img` binary that's already included in `virt-launcher` to create the overlay instead?

Acedus replied:
Ack, guess we'll re-examine this once this https://issues.redhat.com/browse/RHEL-73509 makes it in.

ShellyKa13 commented on pkg/virt-launcher/virtwrap/converter/converter.go:
```diff
@@ -2020,11 +2088,15 @@ func newDeviceNamer(volumeStatuses []v1.VolumeStatus, disks []v1.Disk) map[strin
 	return prefixMap
 }
 
+func GetVolumeNameByDisk(disk api.Disk) string {
```
yeah.. its not used in this PR eventually.. will be used in the next.. but makes the code a bit cleaner anyways

mhenriks commented:
@ShellyKa13 Looks pretty good from initial glance, but I'm not sure if the functionality here is sufficient to merge on its own. Especially without a featuregate

ShellyKa13 commented:
Thanks @mhenriks Totally get the concern!
Yeah, this PR isn’t super valuable on its own, but it also doesn’t really do any harm. If we try to bundle in all the backup logic here, it’ll turn into a giant PR that’s going to be tough to review and maintain.
I can add a feature gate for the incremental backup in this one—so even if someone updates the CR with matching labels, CBT won’t actually kick in unless the gate is enabled. Would that be enough to make it safe to merge now, or do you think it’s still better to hold off and bundle everything into one big PR?
I’m fine either way—just trying to keep things manageable (and make life easier for reviewers lol).

Acedus reviewed (commented):
Looks great! Left a few very small comments.
Still not lgtming as I'm waiting on the resolution of @alicefr's and @mhenriks's comments.

ShellyKa13 commented on pkg/storage/backup/cbt.go:
```diff
@@ -35,21 +37,29 @@ import (
-		return false
+		return false, nil
+	}
+	if !clusterConfig.IncrementalBackupEnabled() {
```
which webhook is that? I can add it to update kubevirt admitter, in the create webhook it cant be done.
Ill add a reject in the kubevirt update admitter if the fg is not enabled and the labelselector is updated

mhenriks replied:
yeah the kubevirt webhook but the inability to do anything about "create" is definitely a shortcoming to consider. Don't want to reject all further updates if the field was set on create. Not sure how to best handle that

ShellyKa13 commented on staging/src/kubevirt.io/api/core/v1/types.go:
```diff
@@ -1804,6 +1809,11 @@ type VirtualMachineStatus struct {
+	// ChangedBlockTracking indicates the state of the changedBlockTracking enablement
+	// +nullable
+	// +optional
+	ChangedBlockTracking ChangedBlockTrackingState `json:"changedBlockTracking,omitempty" optional:"true"`
```
in the vm no, and in vmi I add backupStatus but it is added and deleted every backup so I dont think need to group with this field

ShellyKa13 replied:
I get it, though the cbt can possibly be used for other needs other then backup, so not sure if they should be at the same group..

vladikr commented:
> > Thanks @mhenriks Totally get the concern! Yeah, this PR isn’t super valuable on its own, but it also doesn’t really do any harm. If we try to bundle in all the backup logic here, it’ll turn into a giant PR that’s going to be tough to review and maintain. I can add a feature gate for the incremental backup in this one—so even if someone updates the CR with matching labels, CBT won’t actually kick in unless the gate is enabled. Would that be enough to make it safe to merge now, or do you think it’s still better to hold off and bundle everything into one big PR? I’m fine either way—just trying to keep things manageable (and make life easier for reviewers lol). And of course, I’m doing my best to get all the backup logic in before the feature freeze—appreciate any help from the reviewers to get it wrapped up :)
> 
> It is probably okay if we keep this locked under a feature gate if we are confident that there is no risk to regular operation. And these changes seem pretty isolated. But input from @vladikr or @xpivarc would be appreciated

I'm also concerned that this PR will be approved before the feature freeze without the follow-up PR. I think we should postpone this until after the feature freeze to introduce the complete feature in a single release (and test the entire feature). I mean, merge it for the next release.  Another concern is that if this PR is approved before the freeze, follow-ups could be interpreted as bugs requiring a backport - a situation that is undesirable from my pov for the project.

vladikr commented on pkg/storage/backup/cbt.go:
```diff
@@ -0,0 +1,238 @@
+		return false, ErrIncrementalBackupGateDisabled
+
+	}
+	fmt.Println("3")
```
leftover

ShellyKa13 commented on tests/storage/backup.go:
```diff
@@ -0,0 +1,146 @@
+			Expect(libnamespace.AddLabelToNamespace(virtClient, vm.Namespace, backup.CBTKey, "true")).ToNot(HaveOccurred())
+		}),
+	)
+}))
```
not sure it requires a functional test.. I think UT covers that enough wdyt?

ShellyKa13 commented:
/hold
After some discussions, it was decided to move this feature to the next version, mostly due lack of time to gather all the required content and ensure it's a truly valuable and complete experience for the current release.

ShellyKa13 commented:
/unhold
Will appreciate another look from all

Acedus reviewed (approved):
Looks great, thanks @ShellyKa13!

Acedus commented:
/retest-required

mhenriks reviewed (commented):
Looks pretty good! I have a couple old comments that I think are still relevant/worthy of discussion

ShellyKa13 commented on pkg/storage/backup/cbt.go:
```diff
@@ -0,0 +1,171 @@
+func enableChangedBlockTrackingNoVMI(vm *v1.VirtualMachine) {
+	switch vm.Status.ChangedBlockTracking {
+	case v1.ChangedBlockTrackingUndefined, v1.ChangedBlockTrackingPendingRestart, v1.ChangedBlockTrackingInitializing, v1.ChangedBlockTrackingDisabled:
+		vm.Status.ChangedBlockTracking = v1.ChangedBlockTrackingInitializing
```
not sure its worth it, initializing is like pending in my opinion and to support another state for the time the vm is offline seems redundant and has no functional benefit. WDYT?

ShellyKa13 commented on pkg/storage/backup/cbt.go:
```diff
@@ -0,0 +1,239 @@
+
+	obj, exists, err := nsStore.GetByKey(namespace)
+	if err != nil {
+		logger.Warning("Error retrieving vm namespace from informer")
```
you mean why logger.Warning instead of logger.Errorf?

ShellyKa13 commented on pkg/storage/backend-storage/backend-storage.go:
```diff
@@ -536,6 +554,10 @@ func (bs *BackendStorage) createPVC(vmi *corev1.VirtualMachineInstance, labels m
 }
 
+func (bs *BackendStorage) DeletePVCForVMI(vmi *corev1.VirtualMachineInstance, pvcName string) error {
+	return bs.client.CoreV1().PersistentVolumeClaims(vmi.Namespace).Delete(context.Background(), pvcName, metav1.DeleteOptions{})
```
we send a pvcName we got from PVCForVMI in backendstorage which checks for the specific vmi label/prefix in the name of it:
```
func PVCForVMI(pvcStore cache.Store, vmi *corev1.VirtualMachineInstance) *v1.PersistentVolumeClaim {
	var legacyPVC *v1.PersistentVolumeClaim

	objs := pvcStore.List()
	for _, obj := range objs {
		pvc := obj.(*v1.PersistentVolumeClaim)
		if pvc.Namespace != vmi.Namespace {
			continue
		}
		if pvc.DeletionTimestamp != nil {
			continue
		}
		vmName, found := pvc.Labels[PVCPrefix]
		if found && vmName == vmi.Name {
			return pvc
		}
		if pvc.Name == basePVC(vmi) {
			legacyPVC = pvc
		}
	}

	return legacyPVC
}
```

ShellyKa13 commented:
@Acedus @awels @mhenriks I have addressed all the reviews. big change is changing state to be inside a cbtStatus instead of embedded in the vm/vmi spec.
Also adding a note regarding the migration. I have seen that in case the backendstorage is on RWO pvc then the migration fails. I have reported this to libvirt and opened a bug: https://issues.redhat.com/browse/RHEL-113574

mhenriks commented on staging/src/kubevirt.io/api/core/v1/types.go:
```diff
@@ -2042,6 +2052,36 @@ type InstancetypeStatusRef struct {
+// +k8s:openapi-gen=true
+type ChangedBlockTrackingStatus struct {
+	// State represents the current CBT state
+	State ChangedBlockTrackingState `json:"state"`
```
> Once you have it enabled it seems like a redundant condition

To be clear I was suggesting a condition instead of `ChangedBlockTrackingStatus` struct

> Also I think we can put the backup status, like the backup in progress inside the

very cool, how is that reported byt libvirt/qemu?

ShellyKa13 replied:
yeah it was clear, and I still think having it as a condition less suitable.
Regarding the progress nothing that is reported by libvirt or qemu, rather like snapshotinprogress with the name of the backup

mhenriks commented:
/approve

Acedus reviewed (approved):
Looks great, left 2 small nits, feel free to unhold

/hold

ShellyKa13 commented:
/unhold

Acedus commented:
/lgtm
